source/agora/common/BlockStorage.d

```diff
@@ -758,3 +758,253 @@ public size_t readSizeType (File file) @trusted
+            this.hash_idx.insert(
+                HashPosition(
+                    hash_bytes,
+                    block_position
+                )
+            );
+            return true;
+        }
+        catch (Exception ex)
+        {
+            this.writeLog("MemoryStorage.saveBlock: ", ex);
+            return false;
+        }
+    }
+
+    /***************************************************************************
+
+        Read block from the file.
+
+        Params:
+            block = `Block` to read
+            height = height of `Block`
+
+        Returns:
+            Returns true if success, otherwise returns false.
+
+    ***************************************************************************/
+
+    public bool readBlock (ref Block block, size_t height) @safe nothrow
+    {
+        if ((this.height_idx.length == 0) ||
+            (this.height_idx.back.height < height))
+            return false;
+
+        auto finds
+            = this.height_idx[].find!( (a, b) => a.height == b)(height);
+
+        if (finds.empty)
+            return false;
+
+        try
+        {
+            () @trusted {
+                block = cast(Block)this.blocks[finds.front.position];
+            } ();
+            return true;
+        }
+        catch (Exception ex)
+        {
+            this.writeLog("MemoryStorage.readBlock: ", ex);
+            return false;
+        }
+    }
+
+    /***************************************************************************
+
+        Read block from the file.
+
+        Params:
+            block = `Block` to read
+            hash = `Hash` of `Block`
+
+        Returns:
+            Returns true if success, otherwise returns false.
+
+    ***************************************************************************/
+
+    public bool readBlock (ref Block block, Hash hash) @safe nothrow
+    {
+        ubyte[Hash.sizeof] hash_bytes = hash[];
+
+        auto finds
+            = this.hash_idx[].find!((a, b) => a.hash == b)(hash_bytes);
+
+        if (finds.empty)
+            return false;
+
+        try
+        {
+            () @trusted {
+                block = cast(Block)this.blocks[finds.front.position];
+            } ();
+            return true;
+        }
+        catch (Exception ex)
+        {
+            this.writeLog("MemoryStorage.readBlock: ", ex);
+            return false;
+        }
+    }
+
+    /***************************************************************************
+
+        Write error message
+
+        Params:
+            func = function name
+            ex = Instance of `Exception`
+
+    ***************************************************************************/
+
+    private void writeLog (string func, Exception ex) @trusted nothrow
+    {
+        try
+        {
+            stderr.writeln(func, ex.message);
+        }
+        catch (Exception ex)
+        {
+        }
+    }
+}
+
+
```

Comment:
Could you please replace this unittest with the following one? It creates valid blocks with no double-spend transactions:

```D
///
unittest
{
    import agora.common.crypto.Key;
    import agora.consensus.data.Transaction;
    import agora.consensus.Genesis;
    import std.algorithm.comparison;

    const size_t BlockCount = 50;
    MemoryStorage storage = new MemoryStorage();

    const(Block)[] blocks;
    Hash[] block_hashes;

    auto gen_key_pair = getGenesisKeyPair();
    auto gen_block = getGenesisBlock();
    blocks ~= gen_block;
    storage.saveBlock(gen_block);
    block_hashes ~= hashFull(gen_block.header);
    Transaction[] last_txs;

    void genBlocks (size_t count)
    {
        while (--count)
        {
            auto txs = makeChainedTransactions(gen_key_pair, last_txs, 1);
            auto block = makeNewBlock(blocks[$ - 1], txs);
            last_txs = txs;

            blocks ~= block;
            block_hashes ~= hashFull(block.header);
            storage.saveBlock(block);
        }
    }

    genBlocks(BlockCount);

    // load
    Block[] loaded_blocks;
    loaded_blocks.length = BlockCount;
    foreach (idx; 0 .. BlockCount)
        storage.readBlock(loaded_blocks[idx], idx);

    // compare
    assert(equal(blocks, loaded_blocks));

    // test of random access
    import std.random;
    import std.range;

    auto rnd = rndGen;

    Block random_block;
    foreach (height; iota(BlockCount).randomCover(rnd))
    {
        storage.readBlock(random_block, height);
        assert(random_block.header.height == height);
    }

    foreach (idx; iota(BlockCount).randomCover(rnd))
    {
        storage.readBlock(random_block, block_hashes[idx]);
        assert(hashFull(random_block.header) == block_hashes[idx]);
    }
}
```